components/region_cache_memory_engine/src/engine.rs

```diff
@@ -27,31 +28,96 @@ fn cf_to_id(cf: &str) -> usize {
     }
 }
 
+const ENC_KEY_SEQ_LENGTH: usize = std::mem::size_of::<u64>();
+
+#[inline]
+fn encode_key(key: &[u8], seq: u64) -> Bytes {
+    let mut encoded_key = BytesMut::with_capacity(key.len() + ENC_KEY_SEQ_LENGTH);
+    encoded_key.put(key);
+    encoded_key.put_u64(seq);
+    encoded_key.freeze()
+}
+
+#[inline]
+fn decode_key(encoded_key: &[u8]) -> (&[u8], u64) {
+    debug_assert!(encoded_key.len() >= ENC_KEY_SEQ_LENGTH);
+    let seq_offset = encoded_key.len() - ENC_KEY_SEQ_LENGTH;
+    (
+        &encoded_key[..seq_offset],
+        u64::from_be_bytes(
+            encoded_key[seq_offset..seq_offset + ENC_KEY_SEQ_LENGTH]
+                .try_into()
+                .unwrap(),
+        ),
+    )
+}
+
+#[inline]
+fn encode_for_seek_key(key: &[u8], seq: u64) -> Vec<u8> {
```

Comment:
nit: both `BytesMut` and `Vec<u8>` implement `BufMut`

so you could do:
```
#[inline]
fn encode_key(key: &[u8], seq: u64) -> Bytes {
    let e = encode_key_internal::<BytesMut>(key, seq, BytesMut::with_capacity);
    e.freeze()
}

#[inline]
fn encode_key_internal<T: BufMut>(key: &[u8], seq: u64, f: impl FnOnce(usize) -> T) -> T{
    let mut e = f(key.len() + ENC_KEY_SEQ_LENGTH);
    e.put(key);
    e.put_u64(seq);
    e
}

#[inline]
fn encode_seek_key(key: &[u8], seq: u64) -> Vec<u8> {
    encode_key_internal::<Vec<_>>(key, seq, Vec::with_capacity)
}
```

not sure if it matters, though. your call.

Reply:
good suggestion